RetroAchievements/RAWeb
# chore: switch from npm to pnpm

In the Node.js ecosystem, there are three commonly-used package managers: npm, yarn, and pnpm.

There is a bug in npm that keeps requiring me to reset our entire lockfile just to be able to merge changes. This has happened at least three times now: https://github.com/RetroAchievements/RAWeb/pull/2772#issuecomment-2448586914

Most Node.js projects that reach our scale have to decide if continuing to use vanilla npm is right for them, often due to issues like what I've been facing recently. I am strongly of the opinion we will be better-positioned for the future if we switch to pnpm now.

**pnpm is natively supported by Sail and already installed on the build server.**

There are [many](https://pnpm.io/feature-comparison) benefits to using pnpm. For us, the main one is it can automatically resolve lockfile merge conflicts. More information about the motivation behind pnpm's creation can be found [here](https://pnpm.io/motivation).

`api-docs` and `api-js` have already been using pnpm for months. `docs` is still using npm only because I didn't want non-technical contributors to have to worry about one more thing.

**How to test this PR**
Delete your _node_modules_ folder. Then run `sail pnpm install`. pnpm will reinstall _node_modules_ in a more optimized way described in the motivation link above.

All commands should continue working normally. Instead of `npm run x`, do `pnpm x`, eg:
```
pnpm dev
pnpm lint
pnpm tsc
pnpm test
pnpm test:coverage
pnpm build
```
